components/portal/timeout/controllers.js

```diff
@@ -54,13 +67,65 @@ define(['angular'], function(angular) {
       }
     }
 
+    /**
+     * check if the session is close to timing out due to inactivity
+     */
+    function checkInactive() {
+      var inactiveDuration = sessionInactivityService.getInactiveDuration();
+
+      var warningInactiveDuration =
+          toMillis(appInactiveTimeout - (2 * checkInactiveDelay));
+      var dialogInactiveDuration =
+          toMillis(appInactiveTimeout - checkInactiveDelay);
+      var finalInactiveDuration =
+          toMillis(appInactiveTimeout);
+
+      if (inactiveDuration > finalInactiveDuration) {
+        redirect();
+      } else if (inactiveDuration > warningInactiveDuration) {
+        // Trigger dialog with one minute on the clock
+        $timeout(triggerDialog,
+            dialogInactiveDuration - inactiveDuration, false);
+      }
+    }
+
+    /**
+     * Reload to continue the session
+     */
+    function reload() {
+      $window.location.reload(true);
+    }
+
+    /**
+     * Redirect to end the session;
+     */
+    function redirect() {
+      $window.location.replace(MISC_URLS.loginURL);
+    }
+
     /**
      * Trigger a Session Expired dialog box
      */
     function triggerDialog() {
       $mdDialog.show({
-        templateUrl:
-          'portal/timeout/session-expired.html',
+        template:
```

Comment:
Suggest leaving a comment here about why this needs to be an inline template rather than a template externalized to a URL, otherwise it'll be tempting to refactor out this template in the future.